lib/Serge/Engine/Plugin/parse_csv.pm

```diff
@@ -0,0 +1,273 @@
+
+        my %col_for;
+        for(my $i = 0; $i < @$tokens; $i++) {
+            $col_for{lc $tokens->[$i]} = $i;
+        }
+
+        @headers = @$tokens;
+        $csv->combine(@$tokens) or $self->_report_error($csv);
+        $header_line = $csv->string;
+        die "Could not parse the header" if $self->_report_error($csv);
+        push @out_lines, $header_line;
+        \%col_for;
+    };
+
+    my $key_col_idx = $col_for->{$key_col_name};
+    if(!defined($key_col_idx)) {
+        $self->_email_error_and_die(
+            "Couldn't find required field '$key_col_name' in CSV headers "
+            . quoted_list_str([(keys %$col_for)]) . ".");
+    }
+    my $text_col_idx = $col_for->{$text_col_name};
+    if(!defined($text_col_idx)) {
+        $self->_email_error_and_die(
+            "Couldn't find required field '$text_col_name' in CSV headers "
+            . quoted_list_str([(keys %$col_for)]) . ".");
+    }
+    my $context_col_idx = exists $col_for->{$context_col_name} ? $col_for->{$context_col_name} : -1;
+    my $comment_col_idx = exists $col_for->{$comment_col_name} ? $col_for->{$comment_col_name} : -1;
+
+    # When possible, we continue parsing on errors.  We still die at the end,
+    # but this way we'll show the user all errors in the CSV file in one go,
+    # not just the first we encounter.
+    LINE:
+    while(my $tokens = $csv->getline($csv_fh)) {
+        $has_errors++ && next LINE if $self->_report_error($csv);
+
+        if(@$tokens != @headers) {
+            my $line_num = $csv_fh->input_line_number;
+            my $num_tokens = @$tokens;
+            my $num_headers = @headers;
+            $self->_email_error_and_print(
+                "Can't parse line $line_num\n"
+                . "\tError:    The file has $num_headers header fields, but we found $num_tokens tokens at line $line_num.\n"
+                . "\tHeaders:  " . join('|', @headers) . "\n"
+                . "\tContent:  " . join('|', @$tokens));
+            $has_errors = 1;
+            next LINE;
+        }
+
+        my ($key, $text) = @$tokens[$key_col_idx, $text_col_idx];
+        my $context = $context_col_idx >= 0 ? $tokens->[$context_col_idx] : undef;
+        my $comment = $comment_col_idx >= 0 ? $tokens->[$comment_col_idx] : undef;
+
+        if(!$lang || $self->{import_mode}) {
+            &$callbackref($text, $context, $comment, undef, $lang, $key);
+        }
+        else {
+            $tokens->[$text_col_idx] = &$callbackref($text, $context, $comment, undef, $lang, $key);
+            my $status = $csv->combine(@$tokens);
+            $has_errors++ && next LINE if $self->_report_error($csv);
+            push @out_lines, $csv->string;
+        }
+    }
+
+    if(!$csv->eof) {
+        $has_errors++ if $self->_report_error($csv);
+    }
+
+    if($has_errors) {
+        # The errors will be printed or mailed.  This isn't the place to display them.
+        die "Encountered errors.  Cannot continue.";
```

Comment:
No dot needed at the end of the error message; let's not use double spaces between sentences.